java/ql/src/Violations of Best Practice/Undesirable Calls/DoNotCallFinalize.md

```diff
@@ -0,0 +1,59 @@
+## Overview
+
+Triggering garbage collection by directly calling `finalize()` may either have no effect or may trigger unnecessary garbage collection, leading to erratic behavior, performance issues, or deadlock.
+
+## Recommendation
+
+Avoid calling `finalize()` in application code. Allow the JVM to determine a garbage collection schedule instead. If you need to explicitly release resources, provide a specific method to do so, such as by implementing the `AutoCloseable` interface and overriding its `close` method. You can then use a `try-with-resources` block to ensure that the resource is closed.
+
+## Example
+
+```java
+class LocalCache {
+    private Collection<File> cacheFiles = ...;
+    // ...
+}
+
+void main() {
+    LocalCache cache = new LocalCache();
+    // ...
+    cache.finalize(); // NON_COMPLIANT
+}
+
+```
+
+```java
+import java.lang.AutoCloseable;
+import java.lang.Override;
+
+class LocalCache implements AutoCloseable {
+    private Collection<File> cacheFiles = ...;
+    // ...
+
+    @Override
+    public void close() throws Exception {
+        // release resources here if required
+    }
+}
+
+void main() {
+    // COMPLIANT: uses try-with-resources to ensure that
+    // a resource implementing AutoCloseable is closed.
+    try (LocalCache cache = new LocalCache()) {
+        // ...
+    }
+}
+
+```
+
+# Implementation Notes
+
+This rule is focused on the use of existing `finalize()` invocations rather than attempts to write a custom implementation.
```

Comment:
This line is a bit cryptic to me. Is it trying to say "Overrides of `finalize()` should call `super.finalize()`, and that is explicitly excluded from this rule"? If so, I think you should be more explicit about that. It is useful info.

Reply:
Yes, much clearer 👍🏻 .